lib/cloud/clients.go

```diff
@@ -547,6 +562,43 @@ func (c *cloudClients) initAWSSession(region string) (*awssession.Session, error
 	return session, nil
 }
 
+func (c *cloudClients) getAWSSessionForRole(ctx context.Context, cacheKey string, baseSession *awssession.Session, role services.AssumeRole) (*awssession.Session, error) {
+	c.mtx.RLock()
+	if cachedSession, ok := c.awsSessions[cacheKey]; ok {
+		c.mtx.RUnlock()
+		return cachedSession, nil
+	}
+	c.mtx.RUnlock()
+	return c.initAWSSessionForRole(ctx, cacheKey, baseSession, role)
+}
+
+func (c *cloudClients) initAWSSessionForRole(ctx context.Context, cacheKey string, baseSession *awssession.Session, role services.AssumeRole) (*awssession.Session, error) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if cachedSession, ok := c.awsSessions[cacheKey]; ok { // If some other thread already got here first.
+		return cachedSession, nil
+	}
+	logrus.Debugf("Initializing AWS session for assumed role: %q.", cacheKey)
+	// Make a credentials with AssumeRoleProvider and test it out.
+	cred := stscreds.NewCredentials(baseSession, role.RoleARN, func(p *stscreds.AssumeRoleProvider) {
+		if role.ExternalID != "" {
+			p.ExternalID = aws.String(role.ExternalID)
+		}
+	})
+	if _, err := cred.GetWithContext(ctx); err != nil {
+		return nil, trace.Wrap(libcloudaws.ConvertRequestFailureError(err))
+	}
+
+	// Create a new session with the credentials.
+	config := baseSession.Config.Copy().WithCredentials(cred)
+	roleSession, err := awssession.NewSession(config)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	c.awsSessions[cacheKey] = roleSession
```

Comment:
the underlying credentials expire, but each credential in the chain gets refreshed on next use. 

Something I tested was running my DB agent as an assumed role like this:
```
start an SSO session in account 1 on my laptop
  > assume a role in account 1 with env vars
  > db agent starts 
  > agent assumes a role in account 2
```
Then I waited for both role credentials to expire and tried to connect, and this worked as long as my underlying SSO session hadn't expired, which was only a problem because the agent can't refresh that session and it won't be a problem for something like an EC2 instance with an attached identity.